File: Data Structures & Algorithms/same-binary-tree/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        // if(q==null && p==null)
        //     return true;
        // if((p==null || q==null) || (q.val!=p.val))
        //     return false;
        // return isSameTree(p.left,q.left) & isSameTree(p.right,q.right);

        // iterative DFS

        Stack<TreeNode[]> st = new Stack<>();
        st.push(new TreeNode[]{p,q});

        while(!st.isEmpty()){
            TreeNode[] nodes = st.pop();
            if(nodes[0]==null && nodes[1]==null)
                continue;
            if(nodes[0]==null || nodes[1]==null || nodes[0].val!=nodes[1].val)
                return false;
            st.push(new TreeNode[]{nodes[0].left,nodes[1].left});
            st.push(new TreeNode[]{nodes[0].right,nodes[1].right});
        }
        return true;
    }
}

File: Data Structures & Algorithms/path-sum/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False

        stack =[]
        pathSum =0
        stack.append([root,root.val])

        def isLeafNode(root)->bool:
            return not root.left and not root.right

        while stack:
            
            root, pathSum = stack.pop()
            
            if isLeafNode(root) and pathSum == targetSum:
                return True
            
            if root.left:
                stack.append([root.left,pathSum+root.left.val])
            if root.right:
                stack.append([root.right,pathSum+root.right.val])
        return False
            

